src/shogun/classifier/NearestCentroid.cpp

```diff
@@ -38,48 +38,59 @@ namespace shogun{
 		m_is_trained=false;
 	}
 
-
-	bool NearestCentroid::train_machine(std::shared_ptr<Features> data)
+	std::shared_ptr<Machine> NearestCentroid::fit(
+	    std::shared_ptr<Features> feat, std::shared_ptr<Labels> lab)
 	{
-		ASSERT(m_labels)
-		ASSERT(distance)
-		if (data)
-		{
-			if (m_labels->get_num_labels() != data->get_num_vectors())
-				error("Number of training vectors does not match number of labels");
-			distance->init(data, data);
-		}
-		else
-		{
-			data = distance->get_lhs();
-		}
+		// ASSERT(distance)
+		require(distance, "distance not set");
+		require(
+		    lab->get_num_labels() == feat->get_num_vectors(),
+		    "Number of training vectors does not match number of labels");
+		distance->init(feat, feat);
 
-		auto multiclass_labels = m_labels->as<MulticlassLabels>();
-		auto dense_data = data->as<DenseFeatures<float64_t>>();
+		auto multiclass_labels = lab->as<MulticlassLabels>();
+		auto dense_data = feat->as<DenseFeatures<float64_t>>();
 
 		int32_t num_classes = multiclass_labels->get_num_classes();
 		int32_t num_feats = dense_data->get_num_features();
 
 		SGMatrix<float64_t> centroids(num_feats, num_classes);
 		SGVector<int64_t> num_per_class(num_classes);
-
+		linalg::zero(centroids);
+		linalg::zero(num_per_class);
 		auto iter_labels = multiclass_labels->get_labels().begin();
+		std::map<int64_t, int64_t> label_to_index;
+		int64_t index = 0;
+
 		for (const auto& current : DotIterator(dense_data))
 		{
 			const auto current_class = static_cast<int32_t>(*(iter_labels++));
-			auto target = centroids.get_column(current_class);
+			auto curr_index = index;
+			if (label_to_index.find(current_class) == label_to_index.end())
+			{
+				label_to_index[current_class] = index;
+				index_to_labels[index] = current_class;
+				curr_index = index;
+				index = index + 1;
+			}
+			else
+			{
+				curr_index = label_to_index[current_class];
+			}
+			auto target = centroids.get_column(curr_index);
 			current.add(1, target);
-			num_per_class[current_class]++;
+			num_per_class[curr_index]++;
 		}
 
 		SGVector<float64_t> scale(num_classes);
+
 		for (int32_t i=0 ; i<num_classes ; i++)
 		{
 			int32_t total = num_per_class[i];
-			if(total>1)
-				scale[i] = 1.0/((float64_t)(total-1));
+			if (total > 0)
+				scale[i] = 1.0 / (float64_t)(total);
```

Comment:
```suggestion
				scale[i] = 1.0 / static_cast<float64_t>(total);
```
Doesn't change anything but it is more explicit